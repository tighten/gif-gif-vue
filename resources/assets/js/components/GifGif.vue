<template>
    <div>
        <div class="rounded shadow text-grey-darker">
            <div class="py-8">
                <h3 class="text-center">How do you pronounce the word <span class="font-light uppercase">gif</span>?</h3>

                <div class="flex justify-center py-8">
                    <label for="jif" class="mx-8">
                        GIF
                        <input type="radio" id="jif" value="jif" name="pronunciation" v-model="newVote.picked" />
                    </label>

                    <label for="gif" class="mx-8">
                        GIF
                        <input type="radio" id="gif" value="gif" name="pronunciation" v-model="newVote.picked" />
                    </label>
                </div>
            </div>

            <div class="flex justify-center">
                <div class="w-3/4">
                    <input class="input"
                           id="reason"
                           v-model="newVote.reason"
                           type="text"
                           placeholder="Why do you think it should be pronounced that way?" />
                </div>
            </div>

            <div class="flex justify-center my-4">
                <div class="w-3/4">
                    <input class="input"
                           id="name"
                           v-model="newVote.name"
                           type="text"
                           placeholder="What is your name?" />
                </div>
            </div>

            <div class="flex justify-center">
                <button id="submit"
                        @click="add"
                        :disabled="! valid"
                        :class="{'opacity-50 cursor-not-allowed': !valid }"
                        class="button">
                    Submit
                </button>
            </div>
        </div>

        <div class="py-8">
            <h4 class="text-grey-darker text-center">User Responses:</h4>
            <ul v-for="vote in votes">
                <li class="py-4">
                    <p><span class="font-light uppercase">gif</span> â€”
                        {{ vote.reason }}</p>
                    <p class="text-sm mt-2">Submitted by <span class="uppercase">{{ vote.name }}</span></p>
                </li>
            </ul>
        </div>
    </div>
</template>

<script>
    export default {
        props: ['dataVotes'],

        data() {
            return {
                newVote: {
                    picked: '',
                    reason: '',
                    name: '',
                },

                votes: this.dataVotes,
            }
        },

        methods: {
            add() {
                axios.post('/api/votes', this.newVote)
                    .then(() => {
                        this.votes.push(this.newVote);

                        this.newVote = {
                            picked: '',
                            reason: '',
                            name: '',
                        };
                    });
            }
        },

        computed: {
            valid() {
                return this.newVote.picked && this.newVote.reason && this.newVote.name;
            }
        }
    }
</script>
