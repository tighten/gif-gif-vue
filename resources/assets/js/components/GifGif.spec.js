import { shallowMount } from '@vue/test-utils';
import GifGif from './GifGif';

window.axios = {
    post: jest.fn().mockImplementation(() => Promise.resolve([]))
};

describe('GifGif', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(GifGif, {
            propsData: {
                dataVotes: []
            }
        });
    });

    it('renders', () => {
        expect(wrapper.text()).toContain('How do you pronounce the word gif?');
        expect(wrapper).toMatchSnapshot();
    });

    it('renders with props', () => {
        let vote = {
            picked: 'jif',
            reason: 'It is how you pronounce gin',
            name: 'Samantha'
        };

        wrapper = shallowMount(GifGif, {
            propsData: {
                dataVotes: [vote]
            }
        });

        expect(wrapper.text()).toContain(vote.reason);
        expect(wrapper.text()).toContain(vote.name);
        expect(wrapper).toMatchSnapshot();
    });

    it('updates the selected pronuncation', () => {
        wrapper.find('#jif').trigger('click');
        expect(wrapper.vm.newVote.picked).toBe('jif');

        wrapper.find('#gif').trigger('click');
        expect(wrapper.vm.newVote.picked).toBe('gif');
    });

    it('submits a vote', async () => {
        let vote = {
            picked: 'jif',
            reason: 'It is how you pronounce gin',
            name: 'Samantha'
        };

        expect(wrapper.vm.votes.length).toBe(0);

        // Trigger the radio button of pronuncation
        wrapper.find('#jif').trigger('click');
        expect(wrapper.vm.newVote.picked).toBe(vote.picked);

        // Update the reason input
        let reason = wrapper.find('#reason');
        reason.element.value = vote.reason;
        reason.trigger('input');
        expect(wrapper.vm.newVote.reason).toBe(vote.reason);

        // Update the name input
        let name = wrapper.find('#name');
        name.element.value = vote.name;
        name.trigger('input');
        expect(wrapper.vm.newVote.name).toBe(vote.name);

        // Click the submit button
        wrapper.find('#submit').trigger('click');

        await wrapper.vm.$nextTick();

        expect(axios.post).toBeCalledWith('/api/votes', vote);

        // Check that the vote is added to votes
        expect(wrapper.vm.votes.length).toBe(1);

        // Check that the vote is rendered somewhere
        expect(wrapper.text()).toContain(vote.reason);
        expect(wrapper.text()).toContain(vote.name);
        expect(wrapper.vm.newVote.reason).not.toBe(vote.reason);
        expect(wrapper.vm.newVote.name).not.toBe(vote.name);
    });

    it('cannot submit without all fields', () => {
        wrapper.find('#submit').trigger('click');
        expect(wrapper.vm.votes.length).toBe(0);
    });
});
