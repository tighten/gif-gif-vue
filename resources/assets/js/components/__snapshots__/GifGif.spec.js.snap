// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGif renders 1`] = `
<div>
  <div class="rounded shadow text-grey-darker">
    <div class="py-8">
      <h3 class="text-center">How do you pronounce the word <span class="font-light uppercase">gif</span>?</h3>
      <div class="flex justify-center py-8">
        <label for="jif" class="mx-8">
          GIF
          <input type="radio" id="jif" value="jif" name="pronunciation">
        </label>
        <label for="gif" class="mx-8">
          GIF
          <input type="radio" id="gif" value="gif" name="pronunciation">
        </label>
      </div>
    </div>
    <div class="flex justify-center">
      <div class="w-3/4">
        <input id="reason" type="text" placeholder="Why do you think it should be pronounced that way?" class="input">
      </div>
    </div>
    <div class="flex justify-center my-4">
      <div class="w-3/4">
        <input id="name" type="text" placeholder="What is your name?" class="input">
      </div>
    </div>
    <div class="flex justify-center">
      <button id="submit" disabled="disabled" class="button opacity-50 cursor-not-allowed">
        Submit
      </button>
    </div>
  </div>
  <div class="py-8">
    <h4 class="text-grey-darker text-center">User Responses:</h4>
  </div>
</div>
`;

exports[`GifGif renders with props 1`] = `
<div>
  <div class="rounded shadow text-grey-darker">
    <div class="py-8">
      <h3 class="text-center">How do you pronounce the word <span class="font-light uppercase">gif</span>?</h3>
      <div class="flex justify-center py-8">
        <label for="jif" class="mx-8">
          GIF
          <input type="radio" id="jif" value="jif" name="pronunciation">
        </label>
        <label for="gif" class="mx-8">
          GIF
          <input type="radio" id="gif" value="gif" name="pronunciation">
        </label>
      </div>
    </div>
    <div class="flex justify-center">
      <div class="w-3/4">
        <input id="reason" type="text" placeholder="Why do you think it should be pronounced that way?" class="input">
      </div>
    </div>
    <div class="flex justify-center my-4">
      <div class="w-3/4">
        <input id="name" type="text" placeholder="What is your name?" class="input">
      </div>
    </div>
    <div class="flex justify-center">
      <button id="submit" disabled="disabled" class="button opacity-50 cursor-not-allowed">
        Submit
      </button>
    </div>
  </div>
  <div class="py-8">
    <h4 class="text-grey-darker text-center">User Responses:</h4>
    <ul>
      <li class="py-4">
        <p><span class="font-light uppercase">gif</span> â€” It is how you pronounce gin</p>
        <p class="text-sm mt-2">Submitted by <span class="uppercase">Samantha</span></p>
      </li>
    </ul>
  </div>
</div>
`;
